### DISCUSSION TITLE
About the Data of the Tangent Operator

### DISCUSSION THREAD

**Discussion by lwtcas** [thelfer/MFrontGenericInterfaceSupport#77] [Q&A] [answered]
Hi @thelfer,
   I am a newcomer to the MFront and MGIS. I believe these projects are remarkable and outstanding, worth learning and promoting. Thank you for your great work!
   I've learned and tested all the examples in the directory of 'bindings\fortran\tests'. Now I have a problem: how do I get the data of the tangent operator? I've tried the functions "material_data_manager_get_tangent_operator(K, m)" and "behaviour_data_get_tangent_operator(K, d)", and only got an empty array. Would you like to provide some relevant examples?
  Thanks a lot for any help.

**Comment by thelfer** [comment]
Hi @lwtcas,

Thanks for your interest to MFront and MGIS.

Did you explicitely request the computation of the consistent tangent operator ?

When dealing the behaviour data, you need to set the first element of the tangent operator before the call to `integrate`  to select the kind of tangent operator that you want to request. If you let the default value (`0`), then no computation of the tangent operator is performed. Usually, you can use the value `4` to request the consistent tangent operator.

Please find attach a small working example.

The output shall be:

```bash
 ./TangentOperatorTest 
    0.000    6.000    2.000    8.000    4.000    0.000 
    1.000    7.000    3.000    9.000    5.000    1.000 
    2.000    8.000    4.000    0.000    6.000    2.000 
    3.000    9.000    5.000    1.000    7.000    3.000 
    4.000    0.000    6.000    2.000    8.000    4.000 
    5.000    1.000    7.000    3.000    9.000    5.000 
```
which is, as stated in the documentation, the **transpose** of the consistent tangent operator.

Best,

Thomas

[TangentOperator.zip](https://github.com/thelfer/MFrontGenericInterfaceSupport/files/7833737/TangentOperator.zip)

**Reply by lwtcas** [reply]
Thank you very much for your quick reply！

I directly used the file 'tests\Norton.mfront' earlier. If I understand correctly, do you mean `@Brick StandardElastoViscoPlasticity` is not enough for the computation of the consistent tangent operator and I have to declare the operator explicitly? 

Another question. No argument entry was found to set the type of the tangent operator for the function `integrate(ri, d, b)`. Do you have any suggestions?

**Comment by thelfer** [comment, marked as answer]
@lwtcas,

> I directly used the file 'tests\Norton.mfront' earlier. If I understand correctly, do you mean @Brick StandardElastoViscoPlasticity is not enough for the computation of the consistent tangent operator and I have to declare the operator explicitly?

No, the `StandardElastoViscoPlasticity` does implements the computation of the consistent tangent operator.

But the computation of the consistent tangent operator is requested at runtime (you can integrate the behaviour without computing the consistent tangent operator, for explicit computations, FFT schemes, etc...). The trouble was on the `MGIS` side. See my answer and the example:

> When dealing the behaviour data, you need to set the first element of the tangent operator before the call to integrate to select the kind of tangent operator that you want to request. If you let the default value (0), then no computation of the tangent operator is performed. Usually, you can use the value 4 to request the consistent tangent operator.

The important line in the example is:

```fortran
  K(1,1) = 4
```

which is called just before the `integrate` function.

```fortran
  s = integrate(ri, d, b)
```

See also the C++ documentation of the `integrate` function:

```cxx
  /*!
   * \brief integrate the behaviour. The returned value has the following
   * meaning:
   * - -1: integration failed
   * -  0: integration succeeded but results are unreliable
   * -  1: integration succeeded and results are reliable
   *
   * \param[in,out] d: behaviour data
   * \param[in,out] b: behaviour
   *

   * \note: the type of integration to be performed, must be
   * explicitely set in d.K[0], as follows (see the `IntegrationType` enum).
   *
   * If d.K[0] is greater than 50, the speed of sound must be computed.
   *
   * Let Ke be equal to:
   *
   * - d.K[0] - 100 if d.K[0] is greater than 50
   * - d.K[0] otherwise.
   *
   * If Ke is negative, only the prediction operator is computed and
   * no behaviour integration is performed.
   *
   * Ke has the following meaning:
   *
   * - if Ke is lower than -2.5, the tangent operator must be
   *   computed.
   * - if Ke is in [-2.5:-1.5]: the secant operator must be
   *   computed.
   * - if Ke is in [-1.5:-0.5]: the elastic operator must be
   *   computed.
   * - if Ke is in [-0.5:0.5]: the behaviour integration is
   *   performed, but no stiffness matrix.
   * - if Ke is in [0.5:1.5]: the elastic operator must be
   *   computed.
   * - if Ke is in [1.5:2.5]: the secant operator must be
   *   computed.
   * - if Ke is in [2.5:3.5]: the secant operator must be
   *   computed.
   * - if Ke is greater than 3.5, the consistent tangent operator
   *   must be computed.
   */
  int integrate(BehaviourDataView&, const Behaviour&);
```

**Reply by lwtcas** [reply]
@thelfer,
Your approach worked! Thank you very much!
By the way, I found that the arry of the tangent operator will be emptied once the `update` function is called. I hope this tip will be helpful to other newcomers.

**Reply by thelfer** [reply]
@lwtcas. Great. Do not hesitate to make me feed-backs. The `fortran` bindings has currently only a few users, so I would be interested by your experiments. Do not hesitate to tell me more about your work.
Best,
Thomas
